In [1]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import numpy as np

df = pd.read_csv('../../data/GSE218462_raw_counts_GRCh38.p13_NCBI.tsv', sep='\t')
df = df.T
df.columns = df.iloc[0]
df = df[1:]

scaler = StandardScaler()
scaled_data = scaler.fit_transform(df)
scaled_df = pd.DataFrame(scaled_data)

scaled_df.columns = df.columns
scaled_df.index = df.index
unedited = ['GSM6745632', 'GSM6745633', 'GSM6745634', 'GSM6745635', 'GSM6745636', 'GSM6745637']
scaled_df['Edited (1) or Unedited (0)'] = scaled_df.index.map(lambda gene: 0 if gene in unedited else 1)
mechanisms = {
    "BE4": ["GSM6745599", "GSM6745600", "GSM6745601", "GSM6745611", "GSM6745612", "GSM6745613"],
    "ABE8": ["GSM6745602", "GSM6745603", "GSM6745604", "GSM6745614", "GSM6745615", "GSM6745616"],
    "Cas9": ["GSM6745605", "GSM6745606", "GSM6745607", "GSM6745617", "GSM6745618", "GSM6745619"],
    "Utelectro": ["GSM6745608", "GSM6745609", "GSM6745610", "GSM6745620", "GSM6745621", "GSM6745622"],
    "dCas9": ["GSM6745623", "GSM6745624", "GSM6745625"],
    "BE4alone": ["GSM6745626", "GSM6745627", "GSM6745628"],
    "ABE8alone": ["GSM6745629", "GSM6745630", "GSM6745631"],
    "UT": ["GSM6745632", "GSM6745633", "GSM6745634", "GSM6745635", "GSM6745636", "GSM6745637"]
}

# Inverting the dictionary to map gene code to its corresponding key
mechanism_map = {gene: mechanism for mechanism, genes in mechanisms.items() for gene in genes}

scaled_df['editing mechanism'] = scaled_df.index.map(mechanism_map)

# print(scaled_df)

txt_file_path = '../EDA_sj/relevant_genes_1.6_250.txt'
with open(txt_file_path, 'r') as file:
    txt_data = file.read()

soham_gene_ids = txt_data.split(',')

column_names = scaled_df.columns.tolist()

column_names = [str(col) for col in scaled_df.columns.tolist()]
filtered_column_names = [col for col in scaled_df.columns if str(col) in soham_gene_ids]
scaled_df = scaled_df[filtered_column_names + ['Edited (1) or Unedited (0)', 'editing mechanism']]

# Output the filtered DataFrame
print(scaled_df)

GeneID      113219467     57801      9636  100288175  102465434  109623456  \
GSM6745599  -0.158632  1.070933  1.152150   0.499194  -0.342997  -0.697982   
GSM6745600   1.379276  0.767235  0.761592  -0.765431  -0.342997  -0.697982   
GSM6745601   5.355088  3.457126  2.062135   1.706335  -0.342997  -0.697982   
GSM6745602   2.002351  0.897391  0.336385  -0.075635  -0.342997   0.697982   
GSM6745603   0.679553  1.070933  0.673147   0.614160  -0.342997   0.697982   
GSM6745604  -0.213027  1.548171  0.828458  -0.018153  -0.342997  -0.697982   
GSM6745605   0.009500  0.810621 -0.304008   1.016540  -0.342997  -0.697982   
GSM6745606  -0.393521 -0.274013 -0.724960  -0.650465  -0.342997  -0.697982   
GSM6745607  -0.274840  0.289997 -0.605817   0.384228  -0.342997  -0.697982   
GSM6745609  -0.334181  1.070933 -0.549893   1.591370  -0.342997  -0.697982   
GSM6745610  -0.339126  2.329107 -0.600042   0.211779  -0.342997   0.697982   
GSM6745611  -0.299565 -0.577710  0.256450  -0.305567  -0.342997 

In [2]:
tsv_file_path = '../../data/Human.GRCh38.p13.annot.tsv'
tsv_df = pd.read_csv(tsv_file_path, sep='\t')

tsv_df

/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3553: DtypeWarning: Columns (8,9) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


GeneID       Symbol                                 Description  \
0      100287102      DDX11L1  DEAD/H-box helicase 11 like 1 (pseudogene)   
1         653635       WASH7P           WASP family homolog 7, pseudogene   
2      102466751    MIR6859-1                             microRNA 6859-1   
3      107985730  MIR1302-2HG                         MIR1302-2 host gene   
4      100302278    MIR1302-2                             microRNA 1302-2   
...          ...          ...                                         ...   
39371       4541          ND6                NADH dehydrogenase subunit 6   
39372       4556         TRNE                                    tRNA-Glu   
39373       4519         CYTB                                cytochrome b   
39374       4576         TRNT                                    tRNA-Thr   
39375       4571         TRNP                                    tRNA-Pro   

                        Synonyms        GeneType    EnsemblGeneID  Status  \
0                            NaN          pseudo  ENSG00000290825  active   
1                  FAM39F|WASH5P          pseudo              NaN  active   
2                 hsa-mir-6859-1           ncRNA  ENSG00000278267  active   
3                            NaN           ncRNA              NaN  active   
4      MIRN1302-2|hsa-mir-1302-2           ncRNA  ENSG00000284332  active   
...                          ...             ...              ...     ...   
39371                      MTND6  protein-coding              NaN  active   
39372                       MTTE            tRNA              NaN  active   
39373                      MTCYB  protein-coding              NaN  active   
39374                       MTTT            tRNA              NaN  active   
39375                       MTTP            tRNA              NaN  active   

             ChrAcc ChrStart ChrStop Orientation  Length  \
0      NC_000001.11    11874   14409    positive    1652   
1      NC_000001.11    14362   29370    negative    1769   
2      NC_000001.11    17369   17436    negative      68   
3      NC_000001.11    29926   31295    positive     538   
4      NC_000001.11    30366   30503    positive     138   
...             ...      ...     ...         ...     ...   
39371   NC_012920.1    14149   14673    negative     525   
39372   NC_012920.1    14674   14742    negative      69   
39373   NC_012920.1    14747   15887    positive    1141   
39374   NC_012920.1    15888   15953    positive      66   
39375   NC_012920.1    15956   16023    negative      68   

                  GOFunctionID  \
0                          NaN   
1                          NaN   
2                          NaN   
3                          NaN   
4                          NaN   
...                        ...   
39371               GO:0008137   
39372                      NaN   
39373  GO:0008121///GO:0046872   
39374                      NaN   
39375                      NaN   

                                             GOProcessID  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                             GO:0035195   
...                                                  ...   
39371  GO:0006120///GO:0009060///GO:0032981///GO:0035...   
39372                                                NaN   
39373               GO:0006122///GO:0045333///GO:1902600   
39374                                                NaN   
39375                                                NaN   

                                           GOComponentID  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4        

In [3]:
metadata_file_path = "../../data/Human.GRCh38.p13.annot.tsv"
metadata = pd.read_csv(metadata_file_path, sep='\t')

transposed_data = scaled_df.T
transposed_data = transposed_data.reset_index()
transposed_data.columns.values[0] = 'GeneID'


merged_data = transposed_data.merge(metadata, on='GeneID', how='left').set_index('GeneID')
merged_data = merged_data.iloc[:39378]

print("Merged Data (first few rows):")

# Optionally, save to a new file
output_file_path = "merged_gene_expression_with_metadata.csv"
merged_data.to_csv(output_file_path)
print(merged_data.columns)
# print(f"Merged data saved to {output_file_path}")

/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3553: DtypeWarning: Columns (8,9) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Merged Data (first few rows):
Index(['GSM6745599', 'GSM6745600', 'GSM6745601', 'GSM6745602', 'GSM6745603',
       'GSM6745604', 'GSM6745605', 'GSM6745606', 'GSM6745607', 'GSM6745609',
       'GSM6745610', 'GSM6745611', 'GSM6745612', 'GSM6745613', 'GSM6745614',
       'GSM6745615', 'GSM6745616', 'GSM6745617', 'GSM6745618', 'GSM6745619',
       'GSM6745620', 'GSM6745621', 'GSM6745622', 'GSM6745623', 'GSM6745624',
       'GSM6745625', 'GSM6745626', 'GSM6745627', 'GSM6745628', 'GSM6745629',
       'GSM6745630', 'GSM6745631', 'GSM6745632', 'GSM6745633', 'GSM6745634',
       'GSM6745635', 'GSM6745636', 'GSM6745637', 'Symbol', 'Description',
       'Synonyms', 'GeneType', 'EnsemblGeneID', 'Status', 'ChrAcc', 'ChrStart',
       'ChrStop', 'Orientation', 'Length', 'GOFunctionID', 'GOProcessID',
       'GOComponentID', 'GOFunction', 'GOProcess', 'GOComponent'],
      dtype='object')


In [4]:
merged_data = merged_data.iloc[:, :-6]
merged_data = merged_data.drop(columns=['EnsemblGeneID'])
merged_data
output_file_path = "testing.csv"
merged_data.to_csv(output_file_path)

In [5]:
merged_data_nulls = merged_data['Description'].isnull().sum()
merged_data_nulls

145

In [6]:
d = merged_data.copy()

In [7]:
# Delete any genes with missing descriptions
merged_data = merged_data.dropna(subset=['Description'])
print(f"Number of rows: {merged_data.shape[0]}, Number of columns: {merged_data.shape[1]}")
merged_data.reset_index(inplace=True) # Apparently, I had GeneID as the index, so I reset it to a column cuz it's easier to work with
print(merged_data)

Number of rows: 4656, Number of columns: 48
         GeneID GSM6745599 GSM6745600 GSM6745601 GSM6745602 GSM6745603  \
0     113219467  -0.158632   1.379276   5.355088   2.002351   0.679553   
1         57801   1.070933   0.767235   3.457126   0.897391   1.070933   
2          9636    1.15215   0.761592   2.062135   0.336385   0.673147   
3     100288175   0.499194  -0.765431   1.706335  -0.075635    0.61416   
4     102465434  -0.342997  -0.342997  -0.342997  -0.342997  -0.342997   
...         ...        ...        ...        ...        ...        ...   
4651  105377243  -0.222988  -0.222988  -0.222988  -0.222988  -0.222988   
4652  107987359  -0.164399  -0.164399  -0.164399  -0.164399  -0.164399   
4653       4572  -0.135784   0.567824   3.460437    1.42779   2.131398   
4654       4512  -0.237304   1.302546   5.055768   2.327118   0.856884   
4655       4574  -0.162382   1.353526    5.39109   1.941426   0.658735   

     GSM6745604 GSM6745605 GSM6745606 GSM6745607  ...        Symbol

In [8]:
merged_data

GeneID GSM6745599 GSM6745600 GSM6745601 GSM6745602 GSM6745603  \
0     113219467  -0.158632   1.379276   5.355088   2.002351   0.679553   
1         57801   1.070933   0.767235   3.457126   0.897391   1.070933   
2          9636    1.15215   0.761592   2.062135   0.336385   0.673147   
3     100288175   0.499194  -0.765431   1.706335  -0.075635    0.61416   
4     102465434  -0.342997  -0.342997  -0.342997  -0.342997  -0.342997   
...         ...        ...        ...        ...        ...        ...   
4651  105377243  -0.222988  -0.222988  -0.222988  -0.222988  -0.222988   
4652  107987359  -0.164399  -0.164399  -0.164399  -0.164399  -0.164399   
4653       4572  -0.135784   0.567824   3.460437    1.42779   2.131398   
4654       4512  -0.237304   1.302546   5.055768   2.327118   0.856884   
4655       4574  -0.162382   1.353526    5.39109   1.941426   0.658735   

     GSM6745604 GSM6745605 GSM6745606 GSM6745607  ...        Symbol  \
0     -0.213027     0.0095  -0.393521   -0.27484  ...      MIR12136   
1      1.548171   0.810621  -0.274013   0.289997  ...          HES4   
2      0.828458  -0.304008   -0.72496  -0.605817  ...         ISG15   
3     -0.018153    1.01654  -0.650465   0.384228  ...  LOC100288175   
4     -0.342997  -0.342997  -0.342997  -0.342997  ...       MIR6726   
...         ...        ...        ...        ...  ...           ...   
4651  -0.222988  -0.222988  -0.222988  -0.222988  ...       REREP2Y   
4652  -0.164399  -0.164399  -0.164399  -0.164399  ...  LOC107987359   
4653   1.975041   1.349611   0.098752   0.724182  ...          TRNQ   
4654  -0.052044  -0.139682  -0.558824  -0.258147  ...          COX1   
4655   -0.20611  -0.016622  -0.381023  -0.269273  ...         TRNS1   

                                            Description  \
0                                        microRNA 12136   
1                hes family bHLH transcription factor 4   
2                         ISG15 ubiquitin like modifier   
3                          uncharacterized LOC100288175   
4                                         microRNA 6726   
...                                                 ...   
4651  arginine-glutamic acid dipeptide repeats pseud...   
4652                       uncharacterized LOC107987359   
4653                                           tRNA-Gln   
4654                     cytochrome c oxidase subunit I   
4655                                           tRNA-Ser   

                              Synonyms        GeneType  Status        ChrAcc  \
0                                  NaN           ncRNA  active  NC_000001.11   
1                              bHLHb42  protein-coding  active  NC_000001.11   
2     G1P2|IFI15|IMD38|IP17|UCRP|hUCRP  protein-coding  active  NC_000001.11   
3                                  NaN           ncRNA  active  NC_000001.11   
4                         hsa-mir-6726           ncRNA  active  NC_000001.11   
...                                ...             ...     ...           ...   
4651                               NaN          pseudo  active  NC_000024.10   
4652                               NaN           ncRNA  active   NT_167208.1   
4653                              MTTQ            tRNA  active   NC_012920.1   
4654                         COI|MTCO1  protein-coding  active   NC_012920.1   
4655                             MTTS1            tRNA  active   NC_012920.1   

      ChrStart   ChrStop Orientation  Length  
0       632615    632685    negative    71.0  
1       998962   1001052    negative  1920.0  
2      1013497   1014540    positive   637.0  
3      1059734   1066453    positive   974.0  
4      1296110   1296170    negative    61.0  
...        ...       ...         ...     ...  
4651  26196209  26235374    negative  2725.0  
4652     35630     43681    negative   745.0  
4653      4329      4400    negative    72.0  
4654      5904      7445    positive  1542.0  
4655      7446      7514    negative    69.0  

[4656 rows x 49 columns]

In [9]:
expression_data_T = merged_data.T
# Step 1: Reset the column names using the first row
expression_data_T.columns = expression_data_T.iloc[0]
expression_data_T = expression_data_T.iloc[1:].reset_index(drop=True)
expression_data_T = expression_data_T.apply(pd.to_numeric, errors='ignore')
expression_data_T = expression_data_T[:38]
expression_data_T

GeneID 113219467 57801     9636      100288175 102465434 109623456 8510       \
0      -0.158632  1.070933   1.15215  0.499194 -0.342997 -0.697982  0.160668   
1       1.379276  0.767235  0.761592 -0.765431 -0.342997 -0.697982  -1.27589   
2       5.355088  3.457126  2.062135  1.706335 -0.342997 -0.697982  2.435216   
3       2.002351  0.897391  0.336385 -0.075635 -0.342997  0.697982  0.160668   
4       0.679553  1.070933  0.673147   0.61416 -0.342997  0.697982  0.579663   
5      -0.213027  1.548171  0.828458 -0.018153 -0.342997 -0.697982  0.938803   
6         0.0095  0.810621 -0.304008   1.01654 -0.342997 -0.697982  2.016221   
7      -0.393521 -0.274013  -0.72496 -0.650465 -0.342997 -0.697982 -0.797037   
8       -0.27484  0.289997 -0.605817  0.384228 -0.342997 -0.697982   0.45995   
9      -0.334181  1.070933 -0.549893   1.59137 -0.342997 -0.697982   0.63952   
10     -0.339126  2.329107 -0.600042  0.211779 -0.342997  0.697982  1.776794   
11     -0.299565  -0.57771   0.25645 -0.305567 -0.342997 -0.697982 -0.737181   
12     -0.344071 -0.360783  0.478324  0.326745 -0.342997 -0.697982 -0.138615   
13     -0.173467  0.463538  2.571836  1.936267 -0.342997 -0.697982  2.076077   
14     -0.321818 -0.447554 -0.288507  0.211779 -0.342997  0.697982 -0.198472   
15     -0.035006   0.07307  0.994711  2.223682  2.915476  2.093947  1.357799   
16     -0.306983 -0.621095 -0.180306   0.61416 -0.342997  0.697982  0.160668   
17      -0.30451 -0.447554 -0.598826  0.384228 -0.342997 -0.697982 -0.258328   
18     -0.158632 -0.794637 -0.571472  0.844091 -0.342997 -0.697982 -0.437898   
19     -0.410829 -0.968178 -0.901851 -0.765431 -0.342997 -0.697982 -1.395603   
20     -0.358906 -0.838022 -0.748059 -0.765431 -0.342997 -0.697982 -0.856894   
21     -0.336653 -0.534325 -0.706724 -0.592982 -0.342997 -0.697982 -0.378041   
22      -0.26495 -0.360783 -0.581198 -0.248084 -0.342997 -0.697982 -0.497754   
23     -0.282258 -0.707866 -0.499743 -0.305567 -0.342997  2.093947 -0.138615   
24     -0.260005 -0.621095 -0.443819 -0.190601 -0.342997  0.697982  0.160668   
25     -0.252587 -0.447554 -0.531656  2.166199 -0.342997 -0.697982 -0.138615   
26     -0.173467  1.418015  3.757187  0.269262 -0.342997 -0.697982  1.118372   
27     -0.452862 -0.968178 -0.799728 -1.397743 -0.342997  0.697982 -1.515316   
28      -0.26495 -0.707866  0.269823 -0.018153 -0.342997  0.697982  1.118372   
29     -0.324291 -0.404169 -0.448682 -0.592982 -0.342997 -0.697982 -0.797037   
30     -0.388576  -0.57771 -0.379384 -0.248084 -0.342997 -0.697982 -1.335746   
31     -0.200665 -0.490939 -0.138667 -0.133118 -0.342997 -0.697982 -0.078759   
32     -0.195719 -0.794637 -0.593051 -1.512709 -0.342997  2.093947 -0.378041   
33     -0.190774 -0.751251 -0.522234  -1.34026 -0.342997  2.093947 -0.318185   
34     -0.346543 -1.011563 -0.619798 -1.570192 -0.342997 -0.697982  -0.91675   
35     -0.358906 -0.751251 -0.663869 -1.225294  2.915476  0.697982  -0.91675   
36     -0.361378 -0.534325 -0.503086 -1.110328  2.915476  2.093947 -0.797037   
37     -0.344071 -0.274013 -0.636818 -1.167811  2.915476 -0.697982 -0.856894   

GeneID 107984872 112268220 284661     ... 378949    378950    105377240  \
0      -0.602311 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
1      -0.602311 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
2      -0.602311 -0.164399 -0.164399  ...   2.33513 -0.235702 -0.222988   
3      -0.602311 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
4       0.136006 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
5       0.136006 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
6       2.350955 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
7       0.136006 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
8       2.350955 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
9      -0.602311 -0.164399 -0.164399  ... -0.274721 -0.235702 -0.222988   
10     -0.602311 -0.164399 -0.164399  ... -0.274721 -0

In [10]:
expression_cols = merged_data.columns[:39]
#expression_cols
expression_data = merged_data[expression_cols].apply(pd.to_numeric, errors='coerce')
expression_data

GeneID  GSM6745599  GSM6745600  GSM6745601  GSM6745602  GSM6745603  \
0     113219467   -0.158632    1.379276    5.355088    2.002351    0.679553   
1         57801    1.070933    0.767235    3.457126    0.897391    1.070933   
2          9636    1.152150    0.761592    2.062135    0.336385    0.673147   
3     100288175    0.499194   -0.765431    1.706335   -0.075635    0.614160   
4     102465434   -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   
...         ...         ...         ...         ...         ...         ...   
4651  105377243   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   
4652  107987359   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   
4653       4572   -0.135784    0.567824    3.460437    1.427790    2.131398   
4654       4512   -0.237304    1.302546    5.055768    2.327118    0.856884   
4655       4574   -0.162382    1.353526    5.391090    1.941426    0.658735   

      GSM6745604  GSM6745605  GSM6745606  GSM6745607  ...  GSM6745628  \
0      -0.213027    0.009500   -0.393521   -0.274840  ...   -0.264950   
1       1.548171    0.810621   -0.274013    0.289997  ...   -0.707866   
2       0.828458   -0.304008   -0.724960   -0.605817  ...    0.269823   
3      -0.018153    1.016540   -0.650465    0.384228  ...   -0.018153   
4      -0.342997   -0.342997   -0.342997   -0.342997  ...   -0.342997   
...          ...         ...         ...         ...  ...         ...   
4651   -0.222988   -0.222988   -0.222988   -0.222988  ...   -0.222988   
4652   -0.164399   -0.164399   -0.164399   -0.164399  ...   -0.164399   
4653    1.975041    1.349611    0.098752    0.724182  ...   -0.526678   
4654   -0.052044   -0.139682   -0.558824   -0.258147  ...   -0.286306   
4655   -0.206110   -0.016622   -0.381023   -0.269273  ...   -0.274132   

      GSM6745629  GSM6745630  GSM6745631  GSM6745632  GSM6745633  GSM6745634  \
0      -0.324291   -0.388576   -0.200665   -0.195719   -0.190774   -0.346543   
1      -0.404169   -0.577710   -0.490939   -0.794637   -0.751251   -1.011563   
2      -0.448682   -0.379384   -0.138667   -0.593051   -0.522234   -0.619798   
3      -0.592982   -0.248084   -0.133118   -1.512709   -1.340260   -1.570192   
4      -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   
...          ...         ...         ...         ...         ...         ...   
4651   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   
4652   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   
4653   -0.917571   -0.995750    0.020573   -0.057605   -0.448499   -0.839392   
4654   -0.433196   -0.606873   -0.096739   -0.143264    0.174674   -0.536076   
4655   -0.332436   -0.385882   -0.206110   -0.196393   -0.191534   -0.342153   

      GSM6745635  GSM6745636  GSM6745637  
0      -0.358906   -0.361378   -0.344071  
1      -0.751251   -0.534325   -0.274013  
2      -0.663869   -0.503086   -0.636818  
3      -1.225294   -1.110328   -1.167811  
4       2.915476    2.915476    2.915476  
...          ...         ...         ...  
4651   -0.222988   -0.222988   -0.222988  
4652   -0.164399   -0.164399    6.082763  
4653   -0.683035   -0.448499   -0.604856  
4654   -0.518510   -0.269883   -0.352416  
4655   -0.361588   -0.361588   -0.337295  

[4656 rows x 39 columns]

In [11]:
expression_cols = merged_data.columns[1:39]
expression_data = merged_data[expression_cols].apply(pd.to_numeric, errors='coerce')
expression_data

GSM6745599  GSM6745600  GSM6745601  GSM6745602  GSM6745603  GSM6745604  \
0      -0.158632    1.379276    5.355088    2.002351    0.679553   -0.213027   
1       1.070933    0.767235    3.457126    0.897391    1.070933    1.548171   
2       1.152150    0.761592    2.062135    0.336385    0.673147    0.828458   
3       0.499194   -0.765431    1.706335   -0.075635    0.614160   -0.018153   
4      -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   
...          ...         ...         ...         ...         ...         ...   
4651   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   
4652   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   
4653   -0.135784    0.567824    3.460437    1.427790    2.131398    1.975041   
4654   -0.237304    1.302546    5.055768    2.327118    0.856884   -0.052044   
4655   -0.162382    1.353526    5.391090    1.941426    0.658735   -0.206110   

      GSM6745605  GSM6745606  GSM6745607  GSM6745609  ...  GSM6745628  \
0       0.009500   -0.393521   -0.274840   -0.334181  ...   -0.264950   
1       0.810621   -0.274013    0.289997    1.070933  ...   -0.707866   
2      -0.304008   -0.724960   -0.605817   -0.549893  ...    0.269823   
3       1.016540   -0.650465    0.384228    1.591370  ...   -0.018153   
4      -0.342997   -0.342997   -0.342997   -0.342997  ...   -0.342997   
...          ...         ...         ...         ...  ...         ...   
4651   -0.222988   -0.222988   -0.222988   -0.222988  ...   -0.222988   
4652   -0.164399   -0.164399   -0.164399   -0.164399  ...   -0.164399   
4653    1.349611    0.098752    0.724182   -0.604856  ...   -0.526678   
4654   -0.139682   -0.558824   -0.258147   -0.342890  ...   -0.286306   
4655   -0.016622   -0.381023   -0.269273   -0.322719  ...   -0.274132   

      GSM6745629  GSM6745630  GSM6745631  GSM6745632  GSM6745633  GSM6745634  \
0      -0.324291   -0.388576   -0.200665   -0.195719   -0.190774   -0.346543   
1      -0.404169   -0.577710   -0.490939   -0.794637   -0.751251   -1.011563   
2      -0.448682   -0.379384   -0.138667   -0.593051   -0.522234   -0.619798   
3      -0.592982   -0.248084   -0.133118   -1.512709   -1.340260   -1.570192   
4      -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   -0.342997   
...          ...         ...         ...         ...         ...         ...   
4651   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   -0.222988   
4652   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   -0.164399   
4653   -0.917571   -0.995750    0.020573   -0.057605   -0.448499   -0.839392   
4654   -0.433196   -0.606873   -0.096739   -0.143264    0.174674   -0.536076   
4655   -0.332436   -0.385882   -0.206110   -0.196393   -0.191534   -0.342153   

      GSM6745635  GSM6745636  GSM6745637  
0      -0.358906   -0.361378   -0.344071  
1      -0.751251   -0.534325   -0.274013  
2      -0.663869   -0.503086   -0.636818  
3      -1.225294   -1.110328   -1.167811  
4       2.915476    2.915476    2.915476  
...          ...         ...         ...  
4651   -0.222988   -0.222988   -0.222988  
4652   -0.164399   -0.164399    6.082763  
4653   -0.683035   -0.448499   -0.604856  
4654   -0.518510   -0.269883   -0.352416  
4655   -0.361588   -0.361588   -0.337295  

[4656 rows x 38 columns]

In [12]:
pca_target_variance = 0.75
pca_full = PCA()
expression_data_filled = expression_data.apply(lambda row: row.bfill().ffill(), axis=1)
nan_counts = expression_data.isna().sum().sum()
nan_counts

pca_full.fit(expression_data)

# Determine number of components needed to explain 75% variance
cumulative_variance = pca_full.explained_variance_ratio_.cumsum()
n_components_75 = np.argmax(cumulative_variance >= pca_target_variance) + 1

# Apply PCA with the determined number of components
pca_final = PCA(n_components=n_components_75)
pca_result_final = pca_final.fit_transform(expression_data)

In [13]:
# Retrieve top contributing genes for each principal component
components_df = pd.DataFrame(
    pca_final.components_,
    columns=merged_data['GeneID'][:expression_data.shape[1]],
    index=[f"PC{i+1}" for i in range(n_components_75)]
)

In [14]:
pca_scores = expression_data.values.dot(components_df.values.T)
pca_scores_df = pd.DataFrame(
    pca_scores,
    index=expression_data.index,  # Samples from the original data
    columns=components_df.index   # Principal Components
)

pca_scores_df

PC1       PC2       PC3       PC4       PC5       PC6       PC7  \
0    -1.610801 -1.124601  0.564224 -0.734727 -3.618517 -1.642160  3.554775   
1    -2.911410 -1.735713  1.317743 -0.940426 -3.834176 -0.904999  0.622104   
2    -4.055493 -0.467172  0.203044 -0.044119 -0.943726 -0.452800  0.634827   
3    -4.606621 -2.334358  0.495360 -0.121762 -0.091382 -0.929582 -0.551514   
4    -0.867140  2.928064  2.638784  0.253915  0.111597  0.830826  0.030582   
...        ...       ...       ...       ...       ...       ...       ...   
4651 -2.966374  0.410305 -0.243670  0.490874  1.812046 -0.508316  0.086657   
4652  0.995893  2.984163  5.101125  0.046704  0.373640 -1.017776 -0.270476   
4653 -2.687955 -1.520733  0.628312 -1.352734 -4.187140 -0.689558  0.615231   
4654 -2.110323 -0.949373  0.537358 -0.856578 -3.395222 -1.289364  3.614308   
4655 -1.632372 -1.107345  0.566952 -0.725836 -3.606213 -1.656587  3.575595   

           PC8       PC9      PC10      PC11      PC12      PC13      PC14  \
0    -0.319615 -0.576515  0.317907 -0.604048 -0.587636  0.252575 -0.006866   
1     0.069737  0.498821 -0.576462 -0.288624  1.177010  0.433593 -0.064411   
2     0.404158  0.436351 -0.495258 -0.407477 -0.215540  2.820361 -0.206225   
3     0.220000 -0.915020 -0.358099 -0.298688  1.323217 -0.154980 -0.150400   
4     0.395209  0.473722  0.530572  2.942740 -0.820274 -1.966113 -0.055807   
...        ...       ...       ...       ...       ...       ...       ...   
4651  0.370534  0.569015  0.873461 -0.147608 -0.147457 -3.060324 -0.003758   
4652  0.213154  0.096213  0.190008 -0.669202  0.162325  0.022308 -0.026206   
4653  0.815830  0.977837  0.344002 -1.137483 -0.685559 -0.764796  0.648457   
4654 -0.106106 -0.169933  0.747963 -0.688980 -0.455598 -0.002413  0.120606   
4655 -0.327370 -0.579122  0.250257 -0.601749 -0.581293  0.236220 -0.005946   

          PC15  
0    -0.019470  
1     0.567190  
2     0.456676  
3    -0.670888  
4    -1.096644  
...        ...  
4651 -2.012397  
4652 -0.021736  
4653  0.047816  
4654 -0.049604  
4655 -0.037753  

[4656 rows x 15 columns]

In [15]:
identifiers = d.tail(2)
identifiers

GSM6745599 GSM6745600 GSM6745601 GSM6745602  \
GeneID                                                                   
Edited (1) or Unedited (0)          1          1          1          1   
editing mechanism                 BE4        BE4        BE4       ABE8   

                           GSM6745603 GSM6745604 GSM6745605 GSM6745606  \
GeneID                                                                   
Edited (1) or Unedited (0)          1          1          1          1   
editing mechanism                ABE8       ABE8       Cas9       Cas9   

                           GSM6745607 GSM6745609  ... Symbol Description  \
GeneID                                            ...                      
Edited (1) or Unedited (0)          1          1  ...    NaN         NaN   
editing mechanism                Cas9  Utelectro  ...    NaN         NaN   

                           Synonyms GeneType Status ChrAcc ChrStart ChrStop  \
GeneID                                                                        
Edited (1) or Unedited (0)      NaN      NaN    NaN    NaN      NaN     NaN   
editing mechanism               NaN      NaN    NaN    NaN      NaN     NaN   

                           Orientation Length  
GeneID                                         
Edited (1) or Unedited (0)         NaN    NaN  
editing mechanism                  NaN    NaN  

[2 rows x 48 columns]

In [16]:
y = identifiers.head(1).iloc[:, :38]
y

GSM6745599 GSM6745600 GSM6745601 GSM6745602  \
GeneID                                                                   
Edited (1) or Unedited (0)          1          1          1          1   

                           GSM6745603 GSM6745604 GSM6745605 GSM6745606  \
GeneID                                                                   
Edited (1) or Unedited (0)          1          1          1          1   

                           GSM6745607 GSM6745609  ... GSM6745628 GSM6745629  \
GeneID                                            ...                         
Edited (1) or Unedited (0)          1          1  ...          1          1   

                           GSM6745630 GSM6745631 GSM6745632 GSM6745633  \
GeneID                                                                   
Edited (1) or Unedited (0)          1          1          0          0   

                           GSM6745634 GSM6745635 GSM6745636 GSM6745637  
GeneID                                                                  
Edited (1) or Unedited (0)          0          0          0          0  

[1 rows x 38 columns]

In [17]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score

# Prepare the data
# Assuming components_df is a pandas DataFrame (15x38)
X = components_df.T.values  # Transpose to get samples as rows (38x15)
y = identifiers.head(1).iloc[:, :38]
y = np.array(y).flatten()   # Ensure y is a 1D array (38,)
y = np.array(y).astype(int)  # Ensure y is integer type

print(f"X shape: {X.shape}, y shape: {y.shape}")

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train a classifier (Random Forest as an example)
clf = RandomForestClassifier(class_weight = 'balanced', random_state=42)
clf.fit(X_train, y_train)

# Make predictions
y_pred = clf.predict(X_test)

# Evaluate the classifier
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

X shape: (38, 15), y shape: (38,)
Accuracy: 0.75
Classification Report:
               precision    recall  f1-score   support

           0       0.00      0.00      0.00         2
           1       0.75      1.00      0.86         6

    accuracy                           0.75         8
   macro avg       0.38      0.50      0.43         8
weighted avg       0.56      0.75      0.64         8



/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [18]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, accuracy_score
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline

# Prepare the data
X = components_df.T.values

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Create a pipeline with preprocessing, SMOTE, and classifier
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('smote', SMOTE(k_neighbors = 3, random_state=42)), #SMOTE: balances dataset by generating synthetic samples for unedited class
    ('classifier', RandomForestClassifier(random_state=42))
])

# Define parameter grid for GridSearchCV
param_grid = {
    'classifier__n_estimators': [50, 100, 200], # testing across increments
    'classifier__max_depth': [None, 5, 10], # none for full depth (capture all patterns in data), #5 and 10: avoid overfitting
    'classifier__min_samples_split': [2, 5], # 2: max flexibility in splitting nodes, 5: more samples for a split -> reduce overfitting
    'classifier__min_samples_leaf': [1, 2], # single sample (more depth) or 2 samples (reduce overfitting)
    'classifier__class_weight': ['balanced', 'balanced_subsample'] # balanced: adjusts class weights based on sample freqs; subsample: recalculates weights 
                                                                    # per bootstrapped sample
}

# Perform grid search with cross-validation
grid_search = GridSearchCV(
    pipeline,
    param_grid,
    cv=5, # 5-fold cross validation, balance b/w bias and variance 
    scoring='f1_macro', # equal considerations for both classes, averages the F1 scores of each class
    n_jobs=-1,
    verbose=1
)

# Fit the grid search
grid_search.fit(X_train, y_train)

# Print the best parameters and score
print("Best parameters:", grid_search.best_params_)
print("Best cross-validation score:", grid_search.best_score_)

# Make predictions with the best model
y_pred = grid_search.predict(X_test)

# Print detailed evaluation
print("\nClassification Report:\n", classification_report(y_test, y_pred))

# Get feature importances
feature_importances = pd.DataFrame({
    'feature': range(X.shape[1]),
    'importance': grid_search.best_estimator_.named_steps['classifier'].feature_importances_
})
feature_importances = feature_importances.sort_values('importance', ascending=False)

print("\nTop 5 Most Important Features:")
print(feature_importances.head())

# Perform cross-validation to get a more robust estimate of model performance
cv_scores = cross_val_score(grid_search.best_estimator_, X, y, cv=5, scoring='f1_macro')
print("\nCross-validation scores:", cv_scores)
print("Average CV score:", cv_scores.mean())
print("CV score std:", cv_scores.std())

Fitting 5 folds for each of 72 candidates, totalling 360 fits
Best parameters: {'classifier__class_weight': 'balanced', 'classifier__max_depth': None, 'classifier__min_samples_leaf': 1, 'classifier__min_samples_split': 2, 'classifier__n_estimators': 100}
Best cross-validation score: 0.890909090909091

Classification Report:
               precision    recall  f1-score   support

           0       0.00      0.00      0.00         1
           1       0.88      1.00      0.93         7

    accuracy                           0.88         8
   macro avg       0.44      0.50      0.47         8
weighted avg       0.77      0.88      0.82         8


Top 5 Most Important Features:
   feature  importance
1        1    0.310651
9        9    0.117182
7        7    0.100431
2        2    0.093922
5        5    0.064974


/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/envs/dataplus24/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))



Cross-validation scores: [1.         0.46666667 0.79487179 1.         1.        ]
Average CV score: 0.8523076923076923
CV score std: 0.2085458876248424


Performs best when:
- balanced: model adjusts weights inversely proportional to class frequencies, helping address class imbalance.
- max_depth=None: capture all patterns in data
- min_samples_leaf=1 and min_samples_split=2: full depth, max flexibility in splitting nodes
- small number of trees sufficient for strong performance
The cross-validation score of 0.89 indicates the model performed well during training across folds, particularly in capturing the balance between precision and recall.

Precision, Recall, F1-Score (Support = 8 test samples):

- Class 0: Performance for this class is poor (F1=0.00), likely due to having only 1 sample in the test set. This is too small to generalize reliably.
- Class 1: The model handled this majority class well, achieving perfect recall (100%) and strong precision (88%), leading to a high F1-score of 0.93.

Macro-average is low because it treats all classes equally, including underperforming Class 0.
Weighted average accounts for class imbalance, resulting in a higher score (0.82) due to the better performance for the dominant Class 1.

high variability in performance across folds; decent model performance overall; high variability in accuracy b/w classes